v2/binding/to_event.go

```diff
@@ -61,6 +65,22 @@ func ToEvent(ctx context.Context, message MessageReader, transformers ...Transfo
 	return &e, Transformers(transformers).Transform((*EventMessage)(&e), encoder)
 }
 
+// ToEvents translates a Batch Message and corresponding Reader data to a slice of Events.
+// This function returns the Events generated from the body data, or an error that points
+// to the conversion issue.
+func ToEvents(ctx context.Context, message MessageReader, body io.Reader) ([]event.Event, error) {
```

Comment:
Question: can we have a test with a partial invalid event in the batch request body? I would like to understand the semantics of this API, i.e. which states []event.Event can have depending on the input.

Reply:
Ideally `MessageReader` should define `ReadBatch`, even though currently batching mode is only defined by HTTP protocol binding spec. I'm fine with not adding `ReadBatch` but I wonder if it makes sense to match the `ToEvent` signature and do the type switch in the function body. Or have `ToEvents` take a `*Message` instead of `MessageReader`. 

I think it's fine to not initially supporting `transformers` on batched events until someone ask for it.

Reply:
> Ideally `MessageReader` should define `ReadBatch`, even though currently batching mode is only defined by HTTP protocol binding spec.

I started playing with this, and as soon as that interface changes, it requires changes across lots of codebase, so I'm hesitant to make that change. Or maybe that change should be a subsequent PR for a larger architectural change?

> I'm fine with not adding `ReadBatch` but I wonder if it makes sense to match the `ToEvent` signature and do the type switch in the function body. Or have `ToEvents` take a `*Message` instead of `MessageReader`.

Not sure I'm 100% sure I'm following here. Are you saying that there is a way I could get the `io.Reader` from the `MessageReader`? If so, I'd love some direction - it wasn't immediately obvious to me. 

> 
> I think it's fine to not initially supporting `transformers` on batched events until someone ask for it.

👍🏻

Reply:
> Not sure I'm 100% sure I'm following here. Are you saying that there is a way I could get the io.Reader from the MessageReader? If so, I'd love some direction - it wasn't immediately obvious to me.

You can do something like this:

```
switch mt := message.(type) {
case *http.Message:
  body := mt.BodyReader
  // Since Format doesn't support batch Marshalling, and we know it's structured batch json, we'll go direct to the
  // json.UnMarshall(), since that is the best way to support batch operations for now.
  var events []event.Event
  return events, json.NewDecoder(body).Decode(&events)
}
```

Then return an error for all other cases.

Does that make sense?

Reply:
🤔 interesting. Curious - why have the system return an error at runtime, when you could do the check at compile time?

Reply:
`ToEvents` takes a `MessageReader`, which is more general than `http.Message`. In your case you have only one (http) caller but in theory it does not have to (be the only one).

You can either make `ToEvents` to handle only http message, in which case just use `*http.Message` as the second argument type and remove `body`, or make it a bit more generally applicable and consistent with `ToEvent` and pass MessageReader only.

Reply:
I see, I see 🤔 that makes sense -- we could set ourselves up so if more implementations of Batch come down the pipe, we could handle them here in this place, and expand the switch. SGTM. I'll make the change to `Message` and do the type switch 👍🏻

Reply:
Hit a pretty good wrinkle on making this change. I can't reference the concrete type at all within ToEvents -- because I get into a cyclic dependency issue.

```shell
package github.com/cloudevents/sdk-go/v2
        imports github.com/cloudevents/sdk-go/v2/binding
        imports github.com/cloudevents/sdk-go/v2/protocol/http
        imports github.com/cloudevents/sdk-go/v2/binding: import cycle not allowed
```

So my suggestion - leave this as is, and then as there is a need to create more implementation on Batch processing, implement a `BatchEvents` on `MessageReader` and do the work to implement that across the code base. I figure no point in doing all that work, if the demand may not be there.